chain/network/src/accounts_data/mod.rs

```diff
@@ -62,24 +70,113 @@ pub struct CacheSnapshot {
+    /// just assume that timestamp collision is unlikely).
+    ///
+    /// The alternatives to using a timestamp would be:
+    /// * adding a random_minor_version to AccountData, specifically to avoid collisions
+    ///   (so we would be comparing `(version,random_minor_version)` instead)
+    /// * using some crypto hash function `h` and compare `(version,h(data))`. Assuming that `h`
+    ///   behaves like a random oracle, the semantics will be equivaluent to
+    ///   `random_minor_version`, except that if a node signs exactly the same data and in the
+    ///   previous run, then there will be a collision. But in such a case it doesn't matter
+    ///   since the data is the same.
+    /// * storing `version` of the last signed AccountData in persistent storage,
+    ///   so that the node literally never signs AccountsData with colling versions.
+    ///   This assumption is fragile as long as validators migrate their nodes without copying over
+    ///   the whole storage.
     fn is_new(&self, d: &SignedAccountData) -> bool {
         self.keys.contains(&d.account_key)
             && match self.data.get(&d.account_key) {
-                Some(old) if old.version >= d.version => false,
+                Some(old) if (old.version, old.timestamp) >= (d.version, d.timestamp) => false,
                 _ => true,
             }
     }
 
-    fn try_insert(&mut self, d: Arc<SignedAccountData>) -> Option<Arc<SignedAccountData>> {
+    /// Inserts d into self.data, if
+    /// * `d.account_data` is in self.keys AND
+    /// * `d.version > self.data[d.account_data].version`.
+    /// If d would override local for this node, an AccountData based on `self.local` is signed
+    /// and inserted instead to rollback the overriding change (it can happen in case the node has
+    /// been restarted and we observe the old value emitted by the previous run).
+    /// It returns the newly inserted value (or None if nothing changed).
+    /// The returned value should be broadcasted to the network.
+    fn try_insert(
+        &mut self,
+        clock: &time::Clock,
+        d: Arc<SignedAccountData>,
+    ) -> Option<Arc<SignedAccountData>> {
         if !self.is_new(&d) {
             return None;
         }
+        let d = match &self.local {
+            Some(local) if d.account_key == local.signer.public_key() => Arc::new(
+                VersionedAccountData {
+                    data: local.data.as_ref().clone(),
+                    account_key: local.signer.public_key().clone(),
+                    version: d.version + 1,
+                    timestamp: clock.now_utc(),
+                }
+                .sign(local.signer.as_ref())
+                .unwrap(),
+            ),
+            _ => d,
+        };
         self.data.insert(d.account_key.clone(), d.clone());
         Some(d)
     }
+
+    /// If `self.signer` is in `self.keys` then inserts `d` signed by `self.signer` into self.data
```

Comment:
in these 2 lines you're describing the code and not the intention behind it ;-)

Things I'd mention instead:
- the goal of this function is to set the information about the current node's account
- this function is called only once now (right?) - and in the future it might be called periodically
- if self.key contains the entry for our node - it means that we received such information from the network - and in such case, we want to immediately publish an updated version.

Things I'm missing:
- if we called this function periodically - we'd be publishing updated information every time, right? (as we'd be getting update info from peers all the time)
- we don't check if the account data differs - if it doesn't we don't have to publish any updates.

Reply:
I've clicked submit too early -

Reply:
updated documentation